xmppserver/src/main/java/org/jivesoftware/openfire/muc/spi/LocalMUCUser.java

```diff
@@ -574,6 +575,76 @@ private void processDecliningInvitationMessage(
                                          info.elementTextTrim("reason"), packet.getFrom());
     }
 
+    /**
+     * This method will parse a packet and checks if it is a MEMBER! list retrieval packet or not. 
+     * @param packet IQ
+     * @return true, if it is or false if not
+     */
+    private boolean isListRetrieval(IQ packet)
+    {
+        Element query = packet.getChildElement();
+        if (query!=null&&query.getName().equalsIgnoreCase("query"))
+        {
+            if (query.getNamespace()!=null&&query.getNamespace().getURI().equals("http://jabber.org/protocol/muc#admin"))
+            {
+                Element item = query.element("item");
+                if (item!=null&&item.attributeValue("affiliation")!=null&&
+                    item.attributeValue("affiliation").equalsIgnoreCase("member"))
+                {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Checks if packet sender (non-occupant) is allowed to retrieve a memberlist.
+     * Sender must have a member privileges at minimum to be allowed. 
+     * @param packet IQ
+     * @return true, if sender is allowed to retrieve or false if not
+     */
+    private boolean hasPermissionToRetrieveList(IQ packet)
+    {
+        JID from = packet.getFrom();
+        MUCRoom room = server.hasChatRoom(packet.getTo().getNode())?server.getChatRoom(packet.getTo().getNode()):null;
+        return  room!=null&&(room.getOwners().contains(from.asBareJID())||
```

Comment:
The XEP specifies that 'any occupant of a member-only room' should be allowed. I think it is correct to also allow it for server admins, but maybe it's better to check for occupants, rather than for owners, admins and members? I'm not sure how much work this will be (if it is lots, then don't bother).

Reply:
But if the occupant did not join the room, but he wants the retrieve the list it is senceless to check for occupants whether he es allowed to do i think.
So only allowed users should be able to retrieve.